src/main/java/net/whydah/service/auth/UserAuthenticationResource.java

```diff
@@ -0,0 +1,157 @@
+import net.whydah.util.AdvancedJWTokenUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.RestController;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.FormParam;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.net.URI;
+import java.util.UUID;
+
+import static net.whydah.service.auth.CoreUserResource.API_PATH;
+
+@RestController
+@Path(API_PATH)
+@Produces(MediaType.APPLICATION_JSON)
+public class UserAuthenticationResource extends CoreUserResource {
+    private static final Logger log = LoggerFactory.getLogger(UserAuthenticationResource.class);
+
+    private final SPAApplicationRepository spaApplicationRepository;
+
+    @Autowired
+    public UserAuthenticationResource(CredentialStore credentialStore, SPAApplicationRepository spaApplicationRepository) {
+        super(credentialStore);
+        this.spaApplicationRepository = spaApplicationRepository;
+    }
+
+    @POST
+    @Path("/{secret}/authenticate_user/")
+    public Response authenticateUser(@Context HttpServletRequest httpServletRequest,
+                                     @Context HttpServletResponse httpServletResponse,
+                                     @Context HttpHeaders headers,
+                                     @PathParam("secret") String secret,
+                                     @FormParam("username") String username,
+                                     @FormParam("password") String password) {
+        log.info("Invoked authenticateUser with secret: {} and headers: {}", secret, headers.getRequestHeaders());
+
+        if (username == null || password == null) {
+            log.warn("Unable to find the user credential, returning 403");
+            return Response.status(Response.Status.FORBIDDEN).build();
+        }
+
+        // 1. lookup secret in secret-application session map
+        ApplicationToken applicationToken = spaApplicationRepository.getApplicationTokenBySecret(secret);
+        // 2. execute Whydah API request using the found application
+        if (applicationToken == null) {
+            log.warn("Unable to locate applicationsession from secret, returning 403");
+            return Response.status(Response.Status.FORBIDDEN).build();
+        }
+        Application application = credentialStore.findApplication(applicationToken.getApplicationName());
+
+        UserCredential userCredential = new UserCredential(username, password);
+
+        String ticket = UUID.randomUUID().toString();
+        UserToken userToken = UserTokenMapper.fromUserTokenXml(new CommandLogonUserByUserCredential(
```

Comment:
You could probably split this out to a `getUserToken` method to make the `authenticateUser`-flow more readable

Reply:
Agreed, will split parts into different helper methods. The logic is however not similar across the `authenticateUser` and `getJWTFromTicket` methods, so will need to be multiple helper methods, but it will indeed increase the overall readability!